File: Python/testing-sdk/my_project/test_suites/study_guide_outline/evaluators.py
from typing import List

from autoblocks.testing.models import BaseTestEvaluator
from autoblocks.testing.models import Evaluation
from autoblocks.testing.models import Threshold

from my_project.evaluators.has_substrings import BaseHasSubstrings
from my_project.test_suites.study_guide_outline.test_cases import TestCase


class Formatting(BaseTestEvaluator):
    id = "formatting"

    @staticmethod
    def score(output: str) -> int:
        """
        Every line should either be blank or start with "* " or "  - "
        """
        for line in output.splitlines():
            if not (line.strip() == "" or line.startswith(("* ", "  - "))):
                return 0
        return 1

    def evaluate_test_case(self, test_case: TestCase, output: str) -> Evaluation:
        return Evaluation(score=self.score(output), threshold=Threshold(gte=1))


class NumCategories(BaseTestEvaluator):
    id = "num-categories"

    min_categories: int = 5
    max_categories: int = 10

    def score(self, output: str) -> int:
        return int(self.min_categories <= output.count("* ") <= self.max_categories)

    def evaluate_test_case(self, test_case: TestCase, output: str) -> Evaluation:
        return Evaluation(score=self.score(output), threshold=Threshold(gte=1))


class HasSubstrings(BaseHasSubstrings):
    def expected_substrings(self, test_case: TestCase) -> List[str]:
        return test_case.expected_substrings

File: JavaScript/testing-sdk/src/test-suites/study-guide-outline/evaluators.ts
import { BaseHasSubstrings } from '../../evaluators/has-substrings';
import { BaseTestEvaluator } from '@autoblocks/client/testing';
import type { TestCase } from './test-cases';

export class Formatting extends BaseTestEvaluator<TestCase, string> {
  id = 'formatting';

  /**
   * Every line should either be blank or start with "* " or "  - "
   */
  score(output: string): number {
    for (const line of output.split('\n')) {
      const conditions: boolean[] = [
        line.trim() === '',
        line.startsWith('* '),
        line.startsWith('  - '),
      ];
      if (!conditions.some((c) => c)) {
        return 0;
      }
    }
    return 1;
  }

  evaluateTestCase(args: { testCase: TestCase; output: string }) {
    return {
      score: this.score(args.output),
      threshold: {
        gte: 1,
      },
    };
  }
}

export class NumCategories extends BaseTestEvaluator<TestCase, string> {
  id = 'num-categories';

  minCategories = 5;
  maxCategories = 10;

  score(output: string): number {
    const numCategories = output
      .split('\n')
      .filter((l) => l.startsWith('* ')).length;
    if (
      numCategories >= this.minCategories &&
      numCategories <= this.maxCategories
    ) {
      return 1;
    }
    return 0;
  }

  evaluateTestCase(args: { testCase: TestCase; output: string }) {
    return {
      score: this.score(args.output),
      threshold: {
        gte: 1,
      },
    };
  }
}

export class HasSubstrings extends BaseHasSubstrings<TestCase, string> {
  id = 'has-substrings';

  expectedSubstrings(args: { testCase: TestCase; output: string }): string[] {
    return args.testCase.expectedSubstrings;
  }

  outputAsString(output: string): string {
    return output;
  }
}
